File: src/window_manager.cpp
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/cursorfont.h>
#include <iostream>
#include <vector>
#include "frame.cpp"
using namespace std;

namespace window_manager {
    static bool run = true;
    static Display* display;
    static Window root;
    static Cursor cursor;
    static vector<Frame> frames;

    void map(XMapRequestEvent* event) {
        XMapWindow(display, event->window);
    }
    void unmap(XUnmapEvent* event) {
        XDestroyWindow(display, event->event);
    }

    void setup() {
        if ((display = XOpenDisplay(0)) == 0) {
            cout << "\e[1;31m[ERROR]: FAILED TO OPEN DISPLAY\e[0m" << endl;
            exit(-1);
        }
        if ((root = XDefaultRootWindow(display)) == 0) {
            cout << "\e[1;31m[ERROR]: FAILED TO GET ROOT WINDOW\e[0m" << endl;
            exit(-1);
        }
        if ((cursor = XCreateFontCursor(display, XC_left_ptr)) == 0) {
            cout << "\e[1;31m[ERROR]: FAILED TO CREATE CURSOR\e[0m" << endl;
            exit(-1);
        }
    
        XDefineCursor(display, root, cursor);
        XSelectInput(display, root, SubstructureRedirectMask | SubstructureNotifyMask);
    }
    void loop() {
        XEvent event;
        while (run) {
            XNextEvent(display, &event);
            switch (event.type) {
                case KeyPress:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": KEY PRESSED!\e[0m" << endl;
                    break;
                case ButtonPress:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": BUTTON PRESSED!\e[0m" << endl;
                    break;
                case KeyRelease:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": KEY RELEASED!\e[0m" << endl;
                    break;
                case ButtonRelease:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": BUTTON RELEASED!\e[0m" << endl;
                    break;
                case ConfigureRequest:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": CONFIGURE\e[0m" << endl;
                    break;
                case MapRequest:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": MAP WINDOW\e[0m" << endl;
                    map(&event.xmaprequest);
                    break;
                case MotionNotify:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": MOTION\e[0m" << endl;
                    break;
                case UnmapNotify:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << ": UNMAP WINDOW\e[0m" << endl;
                    unmap(&event.xunmap);
                    break;
                default:
                    cout << "\e[1;32m[LOG]: EVENT: " << event.type << "\e[0m" << endl;
                    break;
            }
        }
        XCloseDisplay(display);
    }
    void kill() {
        run = false;
    }
}

File: src/window_manager.rs
use x11::xlib::{
    self,
    Cursor,
    Display,
    Window,
    XCloseDisplay,
    XCreateFontCursor,
    XDefaultRootWindow,
    XDefineCursor,
    XEvent,
    XNextEvent,
    XOpenDisplay,
    XSelectInput,
    SubstructureNotifyMask,
    SubstructureRedirectMask,
    XMapRequestEvent,
    XMapWindow,
    XUnmapEvent,
    XDestroyWindow,
};
use crate::cursorcont::XC_left_ptr;

static mut RUNNING: bool = true;
static mut DISPLAY: *mut Display = 0 as *mut Display;
static mut ROOTWIN: Window = 0;
static mut CURSOR: Cursor = 0;

pub unsafe fn init() {
    DISPLAY = XOpenDisplay(0 as *const i8);
    if DISPLAY.is_null() {
        panic!("\x1b[1;31m[error] could not open display\x1b[0m");
    }
    ROOTWIN = XDefaultRootWindow(DISPLAY);
    if ROOTWIN == 0 {
        panic!("\x1b[1;31m[error] could not get root window\x1b[0m");
    }
    CURSOR = XCreateFontCursor(DISPLAY, XC_left_ptr);
    if CURSOR == 0 {
        panic!("\x1b[1;31m[error] could not create cursor\x1b[0m");
    }

    XDefineCursor(DISPLAY, ROOTWIN, CURSOR);
    XSelectInput(DISPLAY, ROOTWIN, SubstructureRedirectMask | SubstructureNotifyMask);
}
pub unsafe fn run() {
    let mut event: XEvent = std::mem::zeroed();
    while RUNNING {
        XNextEvent(DISPLAY, &mut event);
        let eventtype = event.get_type();

        if eventtype == xlib::KeyPress{
            println!("\x1b[1;34m[log] key press\x1b[0m");
        } else if eventtype == xlib::ButtonPress{
            println!("\x1b[1;34m[log] button press\x1b[0m");
        } else if eventtype == xlib::KeyRelease{
            println!("\x1b[1;34m[log] key release\x1b[0m");
        } else if eventtype == xlib::ButtonRelease{
            println!("\x1b[1;34m[log] button release\x1b[0m");
        } else if eventtype == xlib::ConfigureRequest{
            println!("\x1b[1;34m[log] configure request\x1b[0m");
        } else if eventtype == xlib::MapRequest{
            println!("\x1b[1;34m[log] map request\x1b[0m");
            map(event.map_request);
        } else if eventtype == xlib::MotionNotify{
            println!("\x1b[1;34m[log] motion notify\x1b[0m");
        } else if eventtype == xlib::UnmapNotify{
            println!("\x1b[1;34m[log] unmap notify\x1b[0m");
            unmap(event.unmap);
        } else {
            println!("\x1b[1;34m[log] unknown event {{ type = {} }}\x1b[0m", eventtype);
        }
    }
    XCloseDisplay(DISPLAY);
}
pub unsafe fn kill() {
    RUNNING = false;
}

unsafe fn map(event: XMapRequestEvent) {
    XMapWindow(DISPLAY, event.window);
}
unsafe fn unmap(event: XUnmapEvent) {
    XDestroyWindow(DISPLAY, event.window);
}
